packages/execute-driver/lib/plugin.js

```diff
@@ -0,0 +1,144 @@
+   * @param {string} [scriptType='webdriverio'] - the name of the driver script
+   * library (currently only webdriverio is supported)
+   *
+   * @returns {Object} - a JSONifiable object representing the return value of
+   * the script
+   */
+  async executeDriverScript (next, driver, script, scriptType = 'webdriverio',
+    timeout = DEFAULT_SCRIPT_TIMEOUT_MS) {
+
+    if (!driver.isFeatureEnabled(FEAT_FLAG)) {
+      throw new Error(`Execute driver script functionality is not available ` +
+                      `unless server is started with --allow-insecure including ` +
+                      `the '${FEAT_FLAG}' flag, e.g., --allow-insecure=${FEAT_FLAG}`);
+    }
+
+    if (scriptType !== SCRIPT_TYPE_WDIO) {
+      throw new TypeError(`Only the '${SCRIPT_TYPE_WDIO}' script type is currently supported`);
+    }
+
+    if (!driver.serverHost || !driver.serverPort) {
+      throw new Error('Address or port of running server were not defined; this ' +
+                      'is required. This is probably a programming error in the driver');
+    }
+
+    if (!_.isNumber(timeout)) {
+      throw new TypeError('Timeout parameter must be a number');
+    }
+
+    const driverOpts = {
+      sessionId: driver.sessionId,
+      // Appium probably won't be behind ssl locally; if it ever is, might need to update this to
+      // provide a user configurable parameter
+      protocol: 'http',
+      hostname: driver.serverHost,
+      port: driver.serverPort,
+      path: driver.serverPath,
+      isW3C: true,
+      isMobile: true,
+      capabilities: driver.caps
+    };
+    this.logger.info(`Constructed webdriverio driver options; W3C mode is ${driverOpts.isW3C ? 'on' : 'off'}`);
+
+
+    // fork the execution script as a child process
+    const childScript = require.resolve('./execute-child.js');
+    this.logger.info(`Forking process to run webdriver script as child using ${childScript}`);
+    const scriptProc = cp.fork(childScript);
+
+    // keep track of whether we have canceled the script timeout, so we can stop
+    // waiting for it and allow this process to finish gracefully
+    let timeoutCanceled = false;
+
+    try {
+      const timer = new timing.Timer();
+      timer.start();
+
+      // promise that deals with the result from the child process
+      const waitForResult = async () => {
+        const res = await new B((res) => {
+          scriptProc.on('message', res); // this is node IPC
+        });
+
+        this.logger.info('Received execute driver script result from child process, shutting it down');
+
+        if (res.error) {
+          throw new Error(res.error.message);
+        }
+
+        return res.success;
+      };
+
+      // promise that waits up to the timeout and throws an error if so, or does
+      // nothing if the timeout is canceled because we got a result from the
+      // child script
+      const waitForTimeout = async () => {
```

Comment:
cannot we simply use B.race here between B.delay and the original script promise?